bindings/matrix-sdk-ffi/src/client.rs

```diff
@@ -385,6 +400,43 @@ impl Client {
     }
 }
 
+pub struct CreateRoomParameters {
+    pub name: String,
+    pub topic: Option<String>,
+    pub is_direct: bool,
+    pub visibility: Visibility,
+    pub preset: create_room::v3::RoomPreset,
+    pub invite: Option<Vec<String>>,
+    pub avatar: Option<String>
+}
+
+impl From<CreateRoomParameters> for create_room::v3::Request  {
+    fn from(value: CreateRoomParameters) -> create_room::v3::Request {
+        let mut request = create_room::v3::Request::new();
+        request.name = Some(value.name);
+        request.topic = value.topic;
+        request.is_direct = value.is_direct;
+        request.visibility = value.visibility;
+        request.preset = Some(value.preset);
+        request.invite = match value.invite {
+            Some(invite) => {
+                let mapped = invite.iter()
+                    .map(UserId::parse)
+                    .collect::<Result<Vec<OwnedUserId>,_>>();
+
+                if let Ok(invite) = mapped { invite } else { vec![] }
+            },
+            None => vec![]
+        };
+        
+        // need a way to inject the avatar url in the request
+        // let avatar = AnyInitialStateEvent::RoomAvatar(...);
```

Comment:
Yes, I think we probably shouldn't even have this trait implementation. To upload the avatar, it needs to be async and have access to the client, so either this conversion should just be a separate `async fn` (not `From` implementation), or inlined into `create_room` entirely.

Does that answer the question?

Reply:
My plan was to call the `upload` from the client and than call the `create_room` passing the url for the state event `m.room.avatar`.

Assuming the client passes the url already, how can I do it at line 433?

Reply:
Oh, now I see. This would be

```rust
let mut content = RoomAvatarEventContent::new();
// content.url = ...
// content.info = ...
let avatar = InitialStateEvent { content, state_key: EmptyStateKey }
let initial_state = vec![Raw::new(avatar).unwrap().cast()];
```

I think `InitialStateEvent` construction could be a little easier, and the `Raw` thing as well. I'll look into it.

Reply:
If you upgrade Ruma to `4eb5033dd0d7c5c6e751fd858bc6763519dbd021` in the main `Cargo.toml`, you can do

```rust
let avatar = InitialStateEvent::new(RoomAvatarEventContent::new());
let initial_state = vec![avatar.to_raw_any()];
```

Reply:
@jplatte I need more help here! 😅

1) The type of `avatar.to_raw_any()` is `Raw<InitialStateEvent<RoomAvatarEventContent>>` while I need a `AnyInitialStateEvent`
2) I'm not getting how I can create an `OwnedMxcUri` from a `String`. In general I'd like to have some clarity around the "Owned" pattern. I mean I often see two versions of the same type eg (`OwnedMxcUri` and `MxcUri`) and I'd like to have more clarity on that.

Reply:
1. That was me being an idiot, https://github.com/ruma/ruma/pull/1513 fixes it
2. The non-owned identifier types from Ruma (where an `Owned` variation exists) are variants `str`s with (usually) extra invariants. `MxcUri` is one of the (two, I think) exceptions in that it doesn't have any extra invariants any can contain an arbitrary string. It's just there to avoid accidental type confusion, and to provide a few convenience methods. To answer your main question, you can use [`OwnedMxcUri::from(string)`](https://docs.ruma.io/ruma/struct.OwnedMxcUri.html#impl-From%3CString%3E-for-OwnedMxcUri) / `string.into()` for the conversion.